scio-extra/src/main/scala/com/spotify/scio/extra/voyager/VoyagerUri.scala

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2023 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.spotify.scio.extra.voyager
+
+import com.spotify.scio.coders.Coder
+import com.spotify.scio.util.{RemoteFileUtil, ScioUtil}
+import com.spotify.voyager.jni.Index
+import com.spotify.voyager.jni.Index.{SpaceType, StorageDataType}
+import org.apache.beam.sdk.options.PipelineOptions
+import org.slf4j.LoggerFactory
+
+import java.io.File
+import java.net.URI
+import java.nio.charset.StandardCharsets
+import java.nio.file.{Files, Path, Paths}
+import scala.collection.mutable
+
+/**
+ * Represents the base URI for a voyager index, either on a local or a remote file system. For
+ * remote file systems, the `path` should be in the form 'scheme://<bucket>/<path>/'. For local
+ * files, it should be in the form '/<path>/'. The `path` specified represents the directory where
+ * the `index.hnsw` and `names.json` are.
+ */
+trait VoyagerUri extends Serializable {
+  val logger = LoggerFactory.getLogger(this.getClass)
```

Comment:
prefer logger declaration in the companion object, with `@transient lazy val` to avoid serialization